subsys/pm/pm.c

```diff
@@ -27,7 +27,36 @@ LOG_MODULE_REGISTER(pm, CONFIG_PM_LOG_LEVEL);
 	(COND_CODE_1(CONFIG_SMP, (arch_curr_cpu()->id), (_current_cpu->id)))
 
 static ATOMIC_DEFINE(z_post_ops_required, CONFIG_MP_MAX_NUM_CPUS);
-static sys_slist_t pm_notifiers = SYS_SLIST_STATIC_INIT(&pm_notifiers);
+
+/* Active state notifiers list */
+static sys_slist_t pm_notifiers_active = SYS_SLIST_STATIC_INIT(&pm_notifiers_active);
+
+/* Generate a notifier list for each DT power state */
+COND_CODE_1(
+	DT_NODE_EXISTS(DT_PATH(cpus, power_states)),
+	(DT_FOREACH_CHILD_STATUS_OKAY_SEP(DT_PATH(cpus, power_states),
+					  Z_PM_NOTIFIERS_CREATE_DT_STATES, (;))),
+	());
+
+/* Array of all notifier lists, active state as first element */
+static sys_slist_t *pm_notifiers_array[] = {
+	&pm_notifiers_active,
+	COND_CODE_1(
+		DT_NODE_EXISTS(DT_PATH(cpus, power_states)),
+		(DT_FOREACH_CHILD_STATUS_OKAY_SEP(DT_PATH(cpus, power_states),
+						  Z_PM_NOTIFIERS_DT_STATES, (,))),
+		())
+};
+
+/* Array of all DT power-states, active state as first element */
+static const struct pm_state_info pmstates[] = {
+	{ .state = PM_STATE_ACTIVE },
+	COND_CODE_1(
+		DT_NODE_EXISTS(DT_PATH(cpus, power_states)),
+		(DT_FOREACH_CHILD_STATUS_OKAY_SEP(DT_PATH(cpus, power_states),
+						  PM_STATE_INFO_DT_INIT, (,))),
+		())
+};
```

Comment:
You don't need it, that is pretty much what you get in `pm_state_cpu_get_all()`.

Reply:
good point about per cpu. I don't see any problem adding this API to return all defined states.
We should just try to avoid duplicate this data, specially in single core cases, that the list is exactly the same of cpu0.

Reply:
Sounds good. To avoid duplication, I could wrap the logic in `#if CONFIG_MP_NUM_CPUS != 1` or similar.